mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
+      if (!operand || operand == reader)
+        continue;
+      if (isa<MemoryEffects::Write>(effect.getEffect())) {
+        onwardWrites.insert(operand);
+        elementwiseArgs.erase(operand);
       }
-      if (isFusorOp(useOp, result)) {
-        fusor = useOp;
-      } else if (auto lgop = dyn_cast<linalg::GenericOp>(useOp)) {
-        if (!fusor && llvm::is_contained(lgop.getOutputs(), result))
-          fusor = useOp;
+      if (isa<MemoryEffects::Read>(effect.getEffect()) &&
+          !onwardWrites.contains(operand)) {
+        elementwiseArgs.insert(operand);
       }
     }
+  }
+  for (OpOperand *arg : elementwiseArgs)
+    collectInputFusionWrites(arg, bufferDeps, state);
+  for (OpOperand *localWrite : onwardWrites) {
+    LLVM_DEBUG(llvm::dbgs() << "Traversing via onward writer: "
+                            << *localWrite->getOwner() << "\n");
+    collectOutputFusionWrites(localWrite, bufferDeps, state);
+  }
+}
 
-    // find fusee's transform chains
-    for (auto &use : buffer.getUses()) {
-      Operation *useOp = use.getOwner();
-      SmallVector<Operation *> chain;
-      bool isWriter = collectChain(buffer, useOp, chain);
-      if (isWriter) {
-        assert(writer == nullptr);
-        writer = useOp;
-      }
-      if (!chain.empty() && chain.back() != fusor) {
-        hasTransforms |= chain.size() > 1;
-        readChains.push_back(chain);
-      }
+////////////////////////////////////////////////////////////////////////
+////  Push Transforms Over alloc to fusor
+////////////////////////////////////////////////////////////////////////
+
+/// In the below, a fusor is the operation that will have some computation fused
+/// into it. This is usually a gemm (either its output or inputs) but can be,
+/// for example, another linalg.generic. The operation that will be fused into
+/// the fusor is the fusee.
+///
+/// A fusor writes to or reads from a memref.alloc() and a fusee does the
+/// opposite of what the fusor does. Before this pass runs, both the fusor and
+/// the fusee may access that intermediate buffer (which will be rewritten away
+/// by fusion) through some sequence of view operations.
+///
+/// After these rewrites, we have the following invariants, which make the
+/// fusion code itself (it's down in AlignTiling.cpp) much simpler and enables
+/// analysis of vectorization within the parts of the pipeline between this pass
+/// and AlignTiling (which can be important for key performance queries).
+///
+/// - The fusee always has a direct reference to the buffer
+///
+/// Furthermore, if multiple fusees have such a non-direct reference, the
+/// intermediate buffer will be duplicated so that each transform stack can be
+/// inverted in isolation.
+
+/// Invert the transform stack between `operand` and `oldAlloc`, replacing all
+/// users but that transform stack with a new allocation whose type is the type
+/// of the front of the transform stack. If the original transforms (and
```

Comment:
The usage of phrase "front of the transform stack" is ambigous and confusing.

I d suggest to reword this to mean: 
1) we invert view stack between operand and oldAlloc
2) create a newAlloc with no views between operand.
3) append the inverted stack to connect rest of the users of oldAlloc to newAlloc